doc/langdef.md

```diff
@@ -116,44 +126,185 @@ In general it is a bad idea for those defining contexts or extensions to use
+
+Maps are a set of key values, and a mapping from these keys to arbitrary values.
+Key values must be an allowed key type: `int`, `uint`, `bool`, or `string`.
+
+Note that the type checker uses a finer-grained notion of list and map types.
+Lists are `list(A)` for the homogenous type `A` of list elements. Maps are
+`map(K, V)` for maps with keys of type `K` and values of type `V`.  The type
+`dyn` is used for heterogeneous values  See
+[Gradual Type Checking](#gradual-type-checking).
+
+Any protocol buffer message is a CEL value, and each message type is its own CEL
+type, represented as its name (e.g. `google.protobuf.timestamp`).
+Some message types will be declared to be _opaque_, which means that CEL
+expressions are not allowed to form literals of this type or to access the
+message fields with the selection operator (see below). Opaque messages can only
+be provided by variables in the evaluation context or returned from functions,
+and they can only be used by providing them as arguments to functions. For
+instance, messages of type `google.protobuf.timestamp` are opaque, and CEL
+expressions will not name their fields directly.
+
+A list can be denoted by the expression `[e1, e2, ..., eN]`, a map by `{ek1:
+ev1, ek2: ev2, ..., ekN: evN}`, and a message by `M{f1: e1, f2: e2, ..., fN: eN}`,
+where `M` must be a simple or qualified name which resolves to a messsage type
+(see [Name Resolution](#resolution)). It is an error to have duplicate keys or
+field names. The empty list, map, and message are `[]`, `{}`, and `M{}`,
+respectively.
 
-Newer versions of protobuf support the well-known types `Struct`, `Value`,
-`Any`, and various wrapper messages for primitives like `Int32Value`. Those
-types have no equivalent in CEL, but instead are automatically converted to one
-of the basic representations introduced in [Values](#values) above as follows:
+See [Field Selection](#field-selection) for accessing elements of lists, maps,
+and messages.
 
-1.  A `Struct` value is converted to an equivalent of `map<String, Value>`.
-2.  A `ListValue` is converted to an equivalent of `repeated Value`.
-3.  A `Value` union is converted to the according value case: double, string,
-    boolean, map, list, or null.
-4.  An `Any` value is on-the-fly deserialized and converted to the message it
-    represents.
-5.  A wrapper is automatically converted into its underlying value, or `null` if
-    the wrapper message is not set.
+### Booleans and Null
 
-NOTE(go/api-expr-open): heterogenous vs homogeneous lists and maps, as well as
-treatment of Any.
+CEL has `true` and `false` as the literals for the `bool` type, with the usual
+meanings.
 
-## Types
+The null value is written `null`. It is used in conversion to and from protocol
+buffer and JSON data, but otherwise has no built-in meaning in CEL. In
+particular, null has its own type (`null_type`) and is not necessarily allowed
+where a value of some other type is expected.
```

Comment:
This paragraph seems to make it unclearer than before what null actually means. What does "not necessarily" allowed mean?

Until we have finished the design of the new null value and type, I guess we better should leave it off the langdef.

Reply:
In other languages, e.g. vanilla Java and C/C++, "null" is allowed where any pointer-ish value is allowed.  In CEL, it might or might not be allowed depending on whether there is an overload for it - we can't make the blanket statement that it's *not* allowed.